A trial for unit test Breadcrumb

DedrickEnc opened the pull request:
I was supposed to send this pull request since Friday, but I had a serious problem of random protractor error. So the main problem for the unit test is essentially finding test cases. So instead of touching directly a big component such as`bhLocation`, I have decided to begin with the small and simplest one `bhBreadcrumb`, I need your suggestions and ideas about my test to improve the bhLocation test on which I am working.

This pull request is testing the `bhBreadcrum` component (unit test)
I have choose to test the component controller particularly his isolated scope (every object, function in the isolated scope).

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [ ] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

DedrickEnc committed:
Initial structure of breadcrumb init test

This commit added an initilal spec.js for the breadcrumb component

jniles commented on client/test/unit/components/bhBreadCrumb.spec.js:
```diff
@@ -0,0 +1,117 @@
+    beforeEach(inject(function ($rootScope, _$componentController_){
+        scope = $rootScope.$new();
+        componentController = _$componentController_;
+        breadCrumb = {
```
This initialization doesn't need to be in the `beforeEach()` call.  It would be faster to take it outside of the function.  Something like this:

``` js
// breadcrumb definition
const breadCrumb = { /* ... */ };

// initialize $injects
beforeEach(inject(($rootScope, _$componentController_) => {
  scope = $rootScope.$new();
  componentController = _$componentController_;
 }));
```

jniles commented on client/test/unit/components/bhBreadCrumb.spec.js:
```diff
@@ -0,0 +1,117 @@
+    });
+
+    beforeEach(inject(function ($rootScope, _$componentController_){
+        scope = $rootScope.$new();
```
I don't see that you use `scope` here.  Do you need it?

jniles commented on client/test/e2e/patient/registry.search.js:
```diff
@@ -115,7 +115,7 @@ function PatientRegistrySearch() {
 
   // combines dates with manual date manipulation
-  it('setting dates manually should find three patients.', function () {
+  it('setting dates manually should find two patients.', function () {
```
:dancer:

jniles commented on client/test/unit/components/bhBreadCrumb.spec.js:
```diff
@@ -0,0 +1,117 @@
+        };
+    }));
+
+    it('receives path data correctly', function (){
```
Since all these tests use the same component (withouth changing initialization parameters), you might move the component initialization step into the `beforeEach()` call.

``` js
let component;

beforeEach(inject(_$componentController_) => {
  component = _$componentController_('bhBreadcrumb', null, angular.copy(breadCrumb));
}));
```

... then, in each test, you can just write ...

``` js
it('receives path data correctly', () => {
  expect(component.path).to.have.length(3);
});
```

jniles commented on client/src/js/components/bhBreadcrumb.js:
```diff
@@ -75,7 +75,7 @@ function BreadcrumbController() {
    * @example
    * vm.bcDropdowns = [
-   *  {
+   *  {\';
```
This should probably be removed before this is merged.  It will break our documentation generation.

jniles commented on client/src/js/components/bhBreadcrumb.js:
```diff
@@ -95,9 +95,12 @@ function BreadcrumbController() {
   vm.bcDropdowns = vm.dropdown || [];
 
-  /** call the appropriate function and update the dropdown label */
+  vm.dropDownItemSelectedLabel = ''; //for the test purpose
```
Hmm.  It usually isn't good practice to add variables into code simply for test purposes, unless there is an easy way to remove them in a `production` build.  Is there any way else to test this component without adding this tag?   Maybe use `child.action()`'s call to trigger an event and detect the event in the test?

DedrickEnc replied:
This function is difficult to test, it is just a way to make it testable, I have tried an other way but no success fill free to help me if you have a good idea.

sfount committed:
Updated test dependencies; verify action links

This commit adds chai-spy as well as a library to preporocess templates
and make them available to the karma test runner. It also introduces
some example tests for verifying that actions passed into a directive
are executed as expected.

DedrickEnc committed:
Uptadating the breadcrumb init test based on review comments

Using chai-spy api to spy function calling
Possibility to compile the HTML string to get a template
Using JQuery to find element and simulate click action
Adding prefix for each ids to make the unique in the page

jniles commented on client/test/unit/components/bhBreadCrumb.spec.js:
```diff
@@ -0,0 +1,119 @@
+      dropdown="config.dropdown">
+    </bh-breadcrumb>`;
+
+  beforeEach(module('pascalprecht.translate', 'ngStorage', 'angularMoment', 'bhima.services', 'bhima.components', 'templates'));
```
Cool. Didn't know you could load multiple modules this way. :+1:

jniles commented on client/src/partials/templates/breadcrumb.tmpl.html:
```diff
@@ -2,10 +2,10 @@
         <span ng-hide="p.link">{{ p.label | translate }}</span>
         <span ng-show="p.link">
-          <a ng-href="{{ p.link }}" href>{{ p.label | translate }}</a>
+          <a ng-href="{{ p.link }}" id="{{ p.id }}" href>{{ p.label | translate }}</a>
```
I would be much more comfortable if this component prefixed the `id`s with some string, like `bc-` or `crumb-btn` or something.  That way, `id`s are guaranteed to be unique on a page.  I can easily imagine a programmer using numeric `id`s that would be used elsewhere in the module template.

sfount replied:
@IMA-WorldHealth/local-contributors This is good practice in the future for component design.

I think this change could be included in this pull request or added as a separate re-factor.

jniles commented on client/test/unit/components/bhBreadCrumb.spec.js:
```diff
@@ -0,0 +1,119 @@
+    scope.$digest();
+
+    scope.config.button.forEach(function (btn) {
+      $(element).find('#' + btn.id).click();
```
If we ever do remove jQuery as a dependency (#176), tests depending on it will break.  It turns out that `angular.element()` is [an alias for the jQuery function](https://docs.angularjs.org/api/ng/function/angular.element) that will default to a smaller implementation if jQuery is not available.  I would advise that we use `angular.element()` instead of jQuery going forward.

From the documentation:

> If jQuery is available, angular.element is an alias for the jQuery function. If jQuery is not available, angular.element delegates to Angular's built-in subset of jQuery, called "jQuery lite" or jqLite.

sfount replied:
Relevant to potentially removing jQuery:
(in this case using a CSS selector with jQuery - potentially not available in jQuery lite) 

From the [angular.element](https://docs.angularjs.org/api/ng/function/angular.element) documentation:

> find() - Limited to lookups by tag name

jniles commented on client/src/partials/templates/breadcrumb.tmpl.html:
```diff
@@ -2,10 +2,10 @@
         <span ng-hide="p.link">{{ p.label | translate }}</span>
         <span ng-show="p.link">
-          <a ng-href="{{ p.link }}" href>{{ p.label | translate }}</a>
+          <a ng-href="{{ p.link }}" id="{{ vm.prefix + p.id }}" href>{{ p.label | translate }}</a>
```
:+1:
